src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -13,193 +12,208 @@ class InMemorySagaPersister : ISagaPersister
     {
         public Task Complete(IContainSagaData sagaData, SynchronizedStorageSession session, ContextBag context)
         {
-            var inMemSession = (InMemorySynchronizedStorageSession) session;
-            inMemSession.Enlist(() =>
-            {
-                VersionedSagaEntity value;
-                if (data.TryRemove(sagaData.Id, out value))
-                {
-                    object lockToken;
-                    lockers.TryRemove(value.LockTokenKey, out lockToken);
-                }
-            });
+            ((InMemorySynchronizedStorageSession)session).Enlist(() =>
+           {
+               var entry = GetEntry(sagaData, context);
+
+               if (sagas.TryRemoveConditionally(sagaData.Id, entry) == false)
+               {
+                   throw new Exception("Saga can't be completed as it was updated by another process.");
+               }
+
+               // saga removed
+               // clean the index
+               if (Equals(entry.CorrelationId, NoCorrelationId) == false)
+               {
+                   byCorrelationId.TryRemoveConditionally(entry.CorrelationId, sagaData.Id);
+               }
+           });
+
             return TaskEx.CompletedTask;
         }
 
-        public Task<TSagaData> Get<TSagaData>(string propertyName, object propertyValue, SynchronizedStorageSession session, ContextBag context) where TSagaData : IContainSagaData
+        public Task<TSagaData> Get<TSagaData>(Guid sagaId, SynchronizedStorageSession session, ContextBag context)
+            where TSagaData : IContainSagaData
         {
-            Guard.AgainstNull(nameof(propertyValue), propertyValue);
+            Entry value;
 
-            foreach (var entity in data.Values)
+            if (sagas.TryGetValue(sagaId, out value))
             {
-                if (!(entity.SagaData is TSagaData))
-                {
-                    continue;
-                }
-
-                var prop = typeof(TSagaData).GetProperty(propertyName);
-                if (prop == null)
-                {
-                    continue;
-                }
-                var existingValue = prop.GetValue(entity.SagaData);
+                context.Set(sagaId.ToString(), value);
```

Comment:
> We can't do it as multiple sagas can be started/be handling the same message.

@Scooletz that's still not true. There is only one saga associated to an incoming message, so no need for a dictionary.
I'm not blocking this PR because of this, but I'm still on the fence regarding that key.